### DISCUSSION TITLE
int4 (4-bit) operations

### DISCUSSION THREAD

**Discussion by andrii0lomakin** [beehive-lab/TornadoVM#467] [Q&A] [answered]
Hello.
I checked the code but decided to double-check.
Does TornadoVM support INT4 (4-bit) arithmetic operations and data types?

**Comment by jjfumero** [comment, marked as answer]
Hi @laa , not at the moment.

**Reply by andrii0lomakin** [reply]
It will probably be much simpler once there is a value class preview. Panama project already has plans to support half-precision float16 arithmetic.

**Reply by jjfumero** [reply]
I do agree. In TornadoVM we took the initiative to support FP16.

https://github.com/beehive-lab/TornadoVM/blob/master/tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/HalfFloatArray.java

So 4-bit int will be similar to FP16, the main issue is the mapping to all target architecture/s that TornadoVM supports. AFAIK, only latest NVIDIA support this.
